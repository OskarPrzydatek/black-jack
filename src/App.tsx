import React from 'react';

import Button from './components/Button';
import CardBox from './components/CardBox';
import CardList from './components/CardList';
import Header from './components/Header';
import Modal from './components/Modal';
import Text from './components/Text';
import { GameActionsEnum } from './constants/gameActions.enum';
import { GameStatusEnum } from './constants/gameStatus.enum';
import { GameActions, GameState } from './models/game.model';
import { gameInitState } from './reducers/gameReducer/gameInitState';
import gameReducer from './reducers/gameReducer/gameReducer';
import * as S from './styles/App.style';
import { GlobalStyle } from './styles/GlobalStyle';

const App: React.FC = () => {
  const [state, dispatch] = React.useReducer<
    React.Reducer<GameState, GameActions>
  >(gameReducer, gameInitState);

  const isNotGameStatusPlay = state.gameStatus !== GameStatusEnum.Play;

  const handleNewGame = () => dispatch({ type: GameActionsEnum.NEW_GAME });

  const handleTakingCard = () => dispatch({ type: GameActionsEnum.GET_A_CARD });

  const handlePass = () => dispatch({ type: GameActionsEnum.PASS });

  return (
    <>
      <S.Main>
        <Header title="Black Jack" />

        <S.GameTabel>
          <CardList dataTestID="player-cards-stack">
            {state.playerStack.map(({ kind, color }) => (
              <CardBox key={`${kind}-${color}`} color={color} kind={kind} />
            ))}
          </CardList>

          <S.GameButtons>
            <Button
              dataTestID="take-card-button"
              isDisabled={isNotGameStatusPlay}
              label="Take Card"
              onClick={handleTakingCard}
            />
            <Button
              dataTestID="pass-button"
              isDisabled={isNotGameStatusPlay}
              label="Pass"
              onClick={handlePass}
            />
          </S.GameButtons>
        </S.GameTabel>

        <Text>Score: {state.gameScore}</Text>

        <Modal
          dataTestID="game-result-modal"
          gameScore={state.gameScore}
          gameStatus={state.gameStatus}
          handleNewGame={handleNewGame}
          newGameButtonTestID="new-game-button"
          resultMessageTestID="game-result-message"
        />
      </S.Main>
      <GlobalStyle />
    </>
  );
};

export default App;
