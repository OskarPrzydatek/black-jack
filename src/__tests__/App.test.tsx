import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import App from '../App';

describe('App', () => {
  test('component snapshot', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('ensure taking card works correctly', () => {
    render(<App />);
    fireEvent.click(screen.getByTestId('take-card-button'));
    const cardStack = screen.getByTestId('player-cards-stack');
    expect(cardStack.childNodes).toHaveLength(2);
  });

  test('ensure pass and new game works correctly', () => {
    render(<App />);
    fireEvent.click(screen.getByTestId('pass-button'));
    expect(screen.queryByTestId('game-result-message')).toHaveTextContent(
      'Pass! End with score: '
    );
    fireEvent.click(screen.getByTestId('new-game-button'));
    expect(screen.queryByTestId('game-result-modal')).not.toBeInTheDocument();
  });
});
