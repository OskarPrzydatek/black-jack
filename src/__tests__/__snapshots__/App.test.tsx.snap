// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component snapshot 1`] = `
<main
  className="sc-iBYQkv hrJZox"
>
  <h1
    className="sc-gKPRtg cRpntk"
  >
    Black Jack
  </h1>
  <div
    className="sc-ftTHYK ytenO"
  >
    <ul
      className="sc-jSUZER hYZLQi"
      data-testid="player-cards-stack"
    >
      <li
        className="sc-gswNZR jVnMTa"
      >
        <div
          className="sc-dkrFOg zlcnh"
        >
          <span
            data-testid="card-box-label"
            style={
              Object {
                "color": "#D91828",
              }
            }
          >
            9
          </span>
        </div>
        <div
          className="sc-eDvSVe jikIfD"
        >
          <svg
            fill="#D91828"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 1c2.21 0 4 1.755 4 3.92C8 2.755 9.79 1 12 1s4 1.755 4 3.92c0 3.263-3.234 4.414-7.608 9.608a.513.513 0 0 1-.784 0C3.234 9.334 0 8.183 0 4.92 0 2.755 1.79 1 4 1z"
            />
          </svg>
        </div>
        <div
          className="sc-dkrFOg sc-hLBbgP zlcnh kxGVxz"
        >
          <span
            style={
              Object {
                "color": "#D91828",
              }
            }
          >
            9
          </span>
        </div>
      </li>
    </ul>
    <div
      className="sc-pyfCe hXJFsd"
    >
      <button
        className="sc-bcXHqe etdubz"
        data-testid="take-card-button"
        disabled={false}
        onClick={[Function]}
      >
        Take Card
      </button>
      <button
        className="sc-bcXHqe etdubz"
        data-testid="pass-button"
        disabled={false}
        onClick={[Function]}
      >
        Pass
      </button>
    </div>
  </div>
  <p>
    Score: 
    0
  </p>
</main>
`;
